### DISCUSSION TITLE
Unable to use --cli-input-yaml with aws cloudformation create-stack

### DISCUSSION THREAD

**Discussion by gautaz** [aws/aws-cli#8866] [Q&A] [answered]
Hello,

I am trying to wrap up arguments for the command `aws cloudformation create-stack` in a YAML file to be then passed with `--cli-input-yaml`.

My current working command line is:

```console
aws cloudformation create-stack --stack-name CfnGitSyncPrerequisites --template-body file://git-sync/prerequisites.yaml --capabilities CAPABILITY_NAMED_IAM
```

Hence reproducing the exact same arguments in a YAML file (`git-sync-prerequisites-stack.yaml`) would from my understanding be with the following content:

```yaml
StackName: CfnGitSyncPrerequisites
TemplateBody: file://git-sync/prerequisites.yaml
Capabilities:
- CAPABILITY_NAMED_IAM
```

But when using this YAML file with `--cli-input-yaml`, it fails:

```console
❯ aws cloudformation create-stack --cli-input-yaml file://git-sync-prerequisites-stack.yaml

An error occurred (ValidationError) when calling the CreateStack operation: Template format error: unsupported structure.
```

I have tried to compare the `--debug` logs of each commands but I was unable to spot the issue (apart that the content of the `git-sync/prerequisites.yaml` file is not displayed in the logs for the `--cli-input-yaml` command).

Any idea on how to debug this or make this work is welcome.

**Comment by uldyssian-sh** [comment, marked as answer]
`--cli-input-yaml` reads an API request structure from your YAML file, but it does not reliably expand `file://...` references inside that YAML into the file contents for `TemplateBody`.

So CloudFormation ends up receiving `TemplateBody` as the literal string `file://git-sync/prerequisites.yaml` (not the template), and then fails template parsing with:

Template format error: unsupported structure.

The two correct ways to make this work are:

Use `--cli-input-yaml` for everything EXCEPT the template body, and pass the template on the command line (CLI args override the input file):

aws cloudformation create-stack \
  --cli-input-yaml file://git-sync-prerequisites-stack.yaml \
  --template-body file://git-sync/prerequisites.yaml


Or, don’t use TemplateBody in the input file at all and use TemplateURL instead (upload the template to S3 and reference it):

StackName: CfnGitSyncPrerequisites
TemplateURL: https://s3.amazonaws.com/<bucket>/git-sync/prerequisites.yaml
Capabilities:
  - CAPABILITY_NAMED_IAM


If you want to debug what is actually being sent, run with `--debug` and look for the serialized request parameters — you will see whether TemplateBody is the real template text or just the literal `file://...` string.

Summary: keep the template file reference on the command line (recommended), or use TemplateURL; embedding `file://...` under TemplateBody inside `--cli-input-yaml` is what triggers the error.

**Reply by uldyssian-sh** [reply]
Thanks! If this helps, please consider marking it as “Helpful” or “Accepted” 😊
